ENH: Allow creating structured void scalars by passing dtype

seberg opened the pull request:
Adds an optional `dtype=` kwarg to `np.void`.  If given (and not None), this kwarg effectively turns it into:

     res = np.array(data, dtype=dtype)[()]

The new dtype argument is keyword-only.

The whole idea behind is that if NEP 51 (scalar repr changes) flies, it would be nice if a new void representation is copy-pastable.  This achieves that.  So it is a small enhancement (I think), that is a stepping stone for fixing scalar representations.

seberg committed:
ENH: Allow creating structured void scalars by passing dtype

Adds an optional `dtype=` kwarg to `np.void`.  If given (and not None),
this kwarg effectively turns it into:

     res = np.array(data, dtype=dtype)[()]

Thanks for Marten's review and Bas' help with the typing.

Reviewed-by: Marten van Kerkwijk <mhvk@astro.utoronto.ca>
Co-authored-by: Bas van Beek <43369155+BvB93@users.noreply.github.com>

mhvk commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -3170,28 +3170,33 @@ static PyObject *
 
-    static char *kwnames[] = {"", NULL};  /* positional-only */
-    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O:void", kwnames, &obj)) {
+    static char *kwnames[] = {"", "dtype", NULL};  /* positional-only */
```
Delete comment `/* positional-only */`

mhvk commented on numpy/core/tests/test_scalar_ctors.py:
```diff
@@ -113,3 +113,57 @@ def test_reals(self, t1, t2):
+    assert res.dtype == "V1"
+    assert res.shape == (18,)
+
+@pytest.mark.parametrize("data",
```
Maybe add a more complicated structure just to be sure? (I know the current code just calls the array creation routine, but might as well have a test that catches bad changes!)
```
def test_void_from_structure():
    data = np.array(((1., 'a'), 2), dtype=[('s', [('f', 'f8'), ('u', 'U1')]), ('i', 'i2')])
    res = np.void(data[()], dtype=data.dtype)
    assert type(res) is np.void
    assert res.dtype == data.dtype
    assert res == data[()]
```

mhvk commented on numpy/core/tests/test_scalar_ctors.py:
```diff
@@ -113,3 +113,57 @@ def test_reals(self, t1, t2):
+            match="void: descr must be a `void.*int64"):
+        np.void(4, dtype="i8")
+
+    dtype = np.dtype("4i")
```
Why is this defined?  I would use '4i' for clarity in the actual test.

seberg replied:
whoops, wanted to avoid a TypeError due to getting it wrong in the test, and then did it wrong :).  Will just match the error string instead.

mhvk reviewed (commented):
Looks good! Only some small points. A more general one is whether keyword-only is needed here. By analogy with `np.array`, I think having the `dtype` as  the second positional argument is OK.

seberg commented:
OK, updated the typing stubs and test (not sure that it is correct) and fixed up those things.

Also allowed it to be passed positionally, it really seems unnecessary to enforce keyword here.  (And added a brief release note.)  I think it should be OK now modulo the typing/typing tests.

seberg commented:
@BvB93 I guess it doesn't work easily to reject certain objects only if `dtype` is not given (or `None`) with the current typing semantics?  It is a bit of a strange overload admittedly...

BvB93 commented on numpy/__init__.pyi:
```diff
@@ -3058,7 +3058,12 @@ class flexible(generic): ...  # type: ignore
 # or a structure
 class void(flexible):
-    def __init__(self, value: _IntLike_co | bytes, /) -> None: ...
+    @overload
+    def __init__(self, value: _IntLike_co | bytes, /, dtype : None = ...) -> None: ...
+    @overload
+    def __init__(self, value: object, /, dtype : None = ...) -> None: NoReturn
+    @overload
+    def __init__(self, value: object, /, dtype: DTypeLike) -> None: ...
```
```suggestion
    @overload
    def __init__(self, value: _IntLike_co | bytes, *, dtype: None = ...) -> None: ...
    @overload
    def __init__(self, value: Any, *, dtype: _DTypeLikeVoid) -> None: ...
```
Note that `_DTypeLikeVoid` still has to be imported from the `numpy._typing` namespace. And if I'm not mistaken `dtype` is keyword-only here, correct?

@seberg the more specific `_DTypeLikeVoid` union (a superset of all acceptable void-based dtype-likes) should help in dealing with this `None`-case, making the `NoReturn` overload redundant.

Secondly as a minor nit: `object` is generally reserved for situations where truly any arbitrary object is accepted (like in the case of `repr` for example), while `Any` moreso describes "not *all* objects are accepted, but I don't have good way to make it more specific".

seberg replied:
Thanks, I had not realized we have that one!

I had dropped the keyword-only, since Marten is right, there is no particular reasons to reject it here.

BvB93 replied:
Ah right, so if it's just a normal keyword-or-positional argument than there's no need for either the `/` or `*`. Might want to update the `np.void` docstring as well though to reflect that.

seberg replied:
Ah, right, thanks!

seberg replied:
Ah, the test failed using `"O,O"` as a void DTypeLike, but I will just change it to a list and consider it a small issue or feature of `_DTypeLikeVoid`

BvB93 replied:
> Ah, the test failed using `"O,O"` as a void DTypeLike

That's a good point actually, I forgot that `_DTypeLikeVoid` only contains specific string-literals in its definition.
Might be better to use `dtype: str | _DTypeLikeVoid` here rather than just plain `_DTypeLikeVoid`.

seberg replied:
OK, you are the typing boss :).

mhvk commented on numpy/typing/tests/data/fail/scalars.pyi:
```diff
@@ -62,7 +63,7 @@ np.uint64(value=0)  # E: Unexpected keyword argument
 np.str_(value='bob')  # E: No overload variant
 np.bytes_(value=b'test')  # E: No overload variant
-np.void(value=b'test')  # E: Unexpected keyword argument
+np.void(value=b'test')  # E: No overload variant
```
Isn't this still an unexpected keyword argument given the `/` in the docstring?

seberg replied:
Yeah, but due to overloads existing, the error changes.  I would agree that the other error text is probably still better :) (although it might miss the note towards overloads).

BvB93 replied:
Agree that the older message is more descriptive. This is a common (minor) problem with overloads, but something that’s also very much an upstream issue.

mhvk reviewed (commented):
Looks all good except possible the typing error.

seberg commented:
I think this is good to go, unless anyone thinks the API addition needs to be announced wider.

mattip commented on numpy/core/_add_newdocs_scalars.py:
```diff
@@ -225,16 +225,52 @@ def add_newdoc_for_scalar_type(obj, fixed_aliases, doc):
+       be returned.
+    dtype : dtype, optional
+        If provided the dtype of the new scalar.  This dtype must
+        be "void" dtype (i.e. a structured or unstructured
```
Can this link to the definition of void https://numpy.org/devdocs/reference/arrays.scalars.html#numpy.void?

seberg replied:
Linking to https://numpy.org/devdocs/user/basics.rec.html#structured-datatypes (lets see if that works), that seems useful.  This is that documentation, so that would be circular ;).

mattip commented on numpy/core/_add_newdocs_scalars.py:
```diff
@@ -225,16 +225,52 @@ def add_newdoc_for_scalar_type(obj, fixed_aliases, doc):
+    has three calling conventions:
+
+    1. ``np.void(5)`` creates a ``dtype="V5"`` scalar filled with
+       ``\0`` bytes.  The 5 can be a Python or NumPy integer.
```
How many ``\0`` bytes?

mattip commented on numpy/core/_add_newdocs_scalars.py:
```diff
@@ -225,16 +225,52 @@ def add_newdoc_for_scalar_type(obj, fixed_aliases, doc):
+    1. ``np.void(5)`` creates a ``dtype="V5"`` scalar filled with
+       ``\0`` bytes.  The 5 can be a Python or NumPy integer.
+    2. ``np.void(b"bytes-like")`` creates a void scalar from
+        the byte string.  The dtype is chosen based on its length.
```
```suggestion
        the byte string.  The actual string dtype is chosen based on its length, in this case `S11`
```

seberg replied:
Adopted somethings similar. (Its V10)

mattip commented on numpy/core/_add_newdocs_scalars.py:
```diff
@@ -225,16 +225,52 @@ def add_newdoc_for_scalar_type(obj, fixed_aliases, doc):
+    2. ``np.void(b"bytes-like")`` creates a void scalar from
+        the byte string.  The dtype is chosen based on its length.
+    3. When a ``dtype=`` is passed the call is rougly the same as an
+       array creation.  However a void scalar is returned when possible.
```
What happens when it is impossible?

seberg replied:
Right, that is not useful info.  Changed it to a scalar being returned.  I probably thought of the fact that I reject subarray dtypes, because we cannot return a scalar for them.

mattip commented on numpy/core/_add_newdocs_scalars.py:
```diff
@@ -225,16 +225,52 @@ def add_newdoc_for_scalar_type(obj, fixed_aliases, doc):
+    >>> np.void((5, 3.2, "eggs"), dtype="i,d,S5")
+    (5, 3.2, b'eggs')  # looks like a tuple, but is `np.void`
+    >>> np.void(3, dtype=[('x', np.int8), ('y', np.int8)])
+    (3, 3)  # looks like a tuple, but is `np.void`
```
Maybe show the dtype in some of these cases?

mattip reviewed (approved):
Just some documentation nits, feel free to mark them as rejected if they don't make sense.

seberg committed:
DOC: Update np.void docs based on Matti's comments

Co-authored-by: Matti Picus <matti.picus@gmail.com>

seberg commented:
Thanks Matti.  Tweaked the docs according to those notes and tests are passing now.  So just going ahead and merging myself.
